casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala

```diff
@@ -0,0 +1,477 @@
+import coop.rchain.comm.transport.TransportLayer
+import coop.rchain.comm.{transport, PeerNode}
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.metrics.Metrics
+import coop.rchain.p2p.effects.PacketHandler
+import coop.rchain.shared.{Log, LogSource, Time}
+import monix.eval.Task
+import monix.execution.Scheduler
+
+import scala.concurrent.duration.FiniteDuration
+import scala.util.Try
+
+object CasperPacketHandler extends CasperPacketHandlerInstances {
+  private implicit val logSource: LogSource = LogSource(this.getClass)
+
+  def apply[F[_]](implicit ev: CasperPacketHandler[F]): CasperPacketHandler[F] = ev
+
+  def of[
+      F[_]: LastApprovedBlock: Metrics: Timer: BlockStore: NodeDiscovery: TransportLayer: ErrorHandler: RPConfAsk: SafetyOracle: Capture: Sync: Time: Monad: Log: MultiParentCasperRef](
+      conf: CasperConf,
+      delay: FiniteDuration,
+      runtimeManager: RuntimeManager,
+      toTask: F[_] => Task[_])(implicit scheduler: Scheduler): F[CasperPacketHandler[F]] =
+    if (conf.approveGenesis) {
+      for {
+        genesis <- Genesis.fromInputFiles[F](conf.bondsFile,
+                                             conf.numValidators,
+                                             conf.genesisPath,
+                                             conf.walletsFile,
+                                             runtimeManager,
+                                             conf.deployTimestamp)
+        _           = println(s"Genesis hash: ${Base16.encode(genesis.blockHash.toByteArray)}")
+        validatorId <- ValidatorIdentity.fromConfig[F](conf)
+        bap         = new BlockApproverProtocol(validatorId.get, genesis, conf.requiredSigs)
+        gv          <- Ref.of[F, CasperPacketHandlerInternal[F]](new GenesisValidatorHandler(bap))
+      } yield new CasperPacketHandlerImpl[F](gv)
+    } else if (conf.createGenesis) {
+      for {
+        genesis <- Genesis.fromInputFiles[F](conf.bondsFile,
+                                             conf.numValidators,
+                                             conf.genesisPath,
+                                             conf.walletsFile,
+                                             runtimeManager,
+                                             conf.deployTimestamp)
+        validatorId      <- ValidatorIdentity.fromConfig[F](conf)
+        bondedValidators <- Genesis.getBondedValidators[F](conf.bondsFile)
+        abp <- ApproveBlockProtocol
+                .of[F](genesis,
+                       bondedValidators,
+                       conf.requiredSigs,
+                       conf.approveGenesisDuration,
+                       conf.approveGenesisInterval)
+                .map(protocol => {
+                  toTask(protocol.run()).forkAndForget.runAsync
+                  protocol
+                })
+        standalone <- Ref.of[F, CasperPacketHandlerInternal[F]](new StandaloneCasperHandler[F](abp))
+        _ <- {
```

Comment:
Should this be in a `Sync.delay` in order to keep the side effect of forking the new process in the `F` context?

Reply:
That is exactly what `Ref.of` is doing: 
```scala
  def of[F[_], A](a: A)(implicit F: Sync[F]): F[Ref[F, A]] = F.delay(unsafe(a))
```

if I wrap it with `Sync[F].delay` then type of standalone is going to be wrapped with `F` effect.

Reply:
That is a good idea.